test/framework/multi_template.go

```diff
@@ -0,0 +1,138 @@
+func (e *ClusterE2ETest) ValidateVsphereMachine(selector string, machineConfig *v1alpha1.VSphereMachineConfig, validations ...VsphereMachineValidation) {
+	e.T.Log("Validating VSphere machine template")
+	ctx := context.Background()
+	machines, err := e.KubectlClient.GetVsphereMachine(ctx, e.Cluster().KubeconfigFile, selector)
+	if err != nil {
+		e.T.Fatal(err)
+	}
+
+	for _, machine := range machines {
+		for _, validation := range validations {
+			err = validation(machineConfig, machine)
+			if err != nil {
+				e.T.Errorf("VSphere machine %v is not valid: %v", machine.Name, err)
+			}
+		}
+	}
+	e.StopIfFailed()
+}
+
+// ValidateWorkerNodeVsphereMachine deduces the worker node group configuration to machine mapping
+// and for each configuration/machine pair executes the provided validation functions.
+func (e *ClusterE2ETest) ValidateWorkerNodeVsphereMachine(validations ...VsphereMachineValidation) {
+	e.T.Log("Validating VSphere worker machine template")
+	machineConfigToMachine := e.getMachineConfigToMachine()
+
+	for wngName, machineName := range machineConfigToMachine {
+		for _, validation := range validations {
+			err := validation(wngName, machineName)
+			if err != nil {
+				e.T.Errorf("VSphere machine %v is not valid: %v", machineName, err)
+			}
+		}
+	}
+	e.StopIfFailed()
+}
+
+func (e *ClusterE2ETest) getMachineConfigToMachine() map[*v1alpha1.VSphereMachineConfig]vspherev1.VSphereMachine {
+	ctx := context.Background()
+	machines, err := e.KubectlClient.GetVsphereMachine(ctx, e.Cluster().KubeconfigFile, workerMachineLabel)
+	if err != nil {
+		e.T.Fatal(err)
+	}
+	wngNameToWng := getWngNameToWng(e.ClusterConfig.Cluster.Spec.WorkerNodeGroupConfigurations)
+
+	machineConfigToMachine := make(map[*v1alpha1.VSphereMachineConfig]vspherev1.VSphereMachine)
+	for _, machine := range machines {
+		if strings.Contains(machine.GetName(), "control-plane") || strings.Contains(machine.Name, "etcd") {
+			continue
+		}
+		if !strings.HasPrefix(machine.GetName(), e.ClusterName) {
+			continue
+		}
+		wng, err := getWngNameFromMachine(machine.GetName(), e.ClusterName)
+		if err != nil {
+			e.T.Fatal(err)
+		}
+		machineConfigName := wngNameToWng[wng].MachineGroupRef.Name
+		machineConfig := e.ClusterConfig.VSphereMachineConfigs[machineConfigName]
+		if _, exists := machineConfigToMachine[machineConfig]; !exists {
+			machineConfigToMachine[machineConfig] = machine
+		}
+	}
+	return machineConfigToMachine
+}
+
+func getWngNameToWng(wngConfigs []v1alpha1.WorkerNodeGroupConfiguration) map[string]v1alpha1.WorkerNodeGroupConfiguration {
+	wngNameToWng := make(map[string]v1alpha1.WorkerNodeGroupConfiguration)
+	for _, wng := range wngConfigs {
+		wngNameToWng[wng.Name] = wng
+	}
+	return wngNameToWng
+}
+
+func getWngNameFromMachine(machineName string, clusterName string) (string, error) {
+	trimmedMachineName := strings.TrimPrefix(machineName, clusterName+"-")
+	wngBaseParts := strings.Split(trimmedMachineName, "-")
+	if len(wngBaseParts) < 3 {
```

Comment:
Add a comment for this function on how we are stripping the string to get the name